ufl/form.py

```diff
@@ -69,7 +69,159 @@ def _sorted_integrals(integrals):
     return tuple(all_integrals)  # integrals_dict
 
 
-class Form(object):
+class BaseForm(object):
+    """Description of an object containing arguments"""
+
+    # Slots is kept empty to enable multiple inheritance with other classes.
+    __slots__ = ()
+    _ufl_is_abstract_ = True
+    _ufl_required_methods_ = ('_analyze_form_arguments', "ufl_domains")
+
+    def __init__(self):
+        # Internal variables for caching form argument data
+        self._arguments = None
+
+    # --- Accessor interface ---
+    def arguments(self):
+        "Return all ``Argument`` objects found in form."
+        if self._arguments is None:
+            self._analyze_form_arguments()
+        return self._arguments
+
```

Comment:
I think `coefficients` can probably be lifted in `BaseForm`, it would make sense for all the other `BaseForm` objects (including `ExternalOperator`).

Reply:
Many `BaseForm` objects can be parametrised by coefficients such as `Form`, `Interp` or `ExternalOperator`. Although there are cases where `coefficients` would just be an empty tuple, it is a general concept that applies to `BaseForm` in the sense that coefficients can be seen as the input of the object in question. So I think we can lift this method to the `BaseForm` class. That's a discussion that took place when making the `BaseForm` class for the first time.